sycl/include/CL/sycl/detail/image_accessor_util.hpp

```diff
@@ -123,85 +188,533 @@ template <typename T> vec<T, 4> readPixel(T *Ptr, image_channel_order Order) {
+    Temp = (cl::sycl::clamp(WriteData, MinVal, MaxVal));
+    PixelData = Temp.convert<ChannelType>();
+    break;
+  case image_channel_type::unsigned_int16:
+    // convert_ushort_sat(Data)
+    MinVal = min_v<cl_ushort>();
+    MaxVal = max_v<cl_ushort>();
+    Temp = cl::sycl::clamp(WriteData, MinVal, MaxVal);
+    PixelData = Temp.convert<ChannelType>();
+    break;
+  case image_channel_type::unsigned_int32:
+    // no conversion is performed.
+    PixelData = WriteData.convert<ChannelType>();
+    break;
+  default:
+    // OpenCL Spec section 6.12.14.4 does not allow writing uint4 data to an
+    // image with channel datatype other than unsigned_int8,unsigned_int16 and
+    // unsigned_int32.
+    throw cl::sycl::invalid_parameter_error(
+        "Datatype of data to write - cl_uint4 is incompatible with the "
+        "image_channel_type of the image.");
+    break;
+  }
   return PixelData;
 }
 
-// Separate function for half datatype.
-// To resolve ambiguity when converting data into half datatype,float is used as
-// typecast.
-template <typename WriteDataType>
-vec<cl_half, 4> convertWriteDataToHalf(vec<WriteDataType, 4> WriteData,
-                                       image_channel_type ImageChannelType) {
-  vec<cl_half, 4> PixelData;
-  PixelData.x() = (float)WriteData.x();
-  PixelData.y() = (float)WriteData.y();
-  PixelData.z() = (float)WriteData.z();
-  PixelData.w() = (float)WriteData.w();
+template <typename ChannelType>
+vec<ChannelType, 4> convertWriteData(vec<cl_int, 4> WriteData,
+                                     image_channel_type ImageChannelType) {
+
+  vec<ChannelType, 4> PixelData;
+  vec<cl_int, 4> Temp;
+  cl_int MinVal;
+  cl_int MaxVal;
+  switch (ImageChannelType) {
+  case image_channel_type::signed_int8:
+    // convert_char_sat(Data)
+    MinVal = min_v<cl_char>();
+    MaxVal = max_v<cl_char>();
+    Temp = cl::sycl::clamp(WriteData, MinVal, MaxVal);
+    PixelData = Temp.convert<ChannelType>();
+    break;
+  case image_channel_type::signed_int16:
+    // convert_short_sat(Data)
+    MinVal = min_v<cl_short>();
+    MaxVal = max_v<cl_short>();
+    Temp = cl::sycl::clamp(WriteData, MinVal, MaxVal);
+    PixelData = Temp.convert<ChannelType>();
+    break;
+  case image_channel_type::signed_int32:
+    PixelData = WriteData.convert<ChannelType>();
+    break;
+  default:
+    // OpenCL Spec section 6.12.14.4 does not allow writing int4 data to an
+    // image with channel datatype other than signed_int8,signed_int16 and
+    // signed_int32.
+    throw cl::sycl::invalid_parameter_error(
+        "Datatype of data to write - cl_int4 is incompatible with the "
+        "image_channel_type of the image.");
+    break;
+  }
+  return PixelData;
+}
+
+// Multiplication operator is not supported in vec class for a float operand.
+// So here each vector element is individually multiplied by the multiplicand.
+template <typename T>
+cl::sycl::vec<T, 4> multiplyEachElement(cl::sycl::vec<T, 4> Operand,
```

Comment:
You can make it general and use for division also:
c++ provides some structs that have overload operator() specified for operations + - * / and etc.
std::plus, std::minus, std::multiplies, std::divides